In [96]:
%matplotlib inline
import numpy as np
import pandas as pd
import scipy
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns
import re
import nltk
import string
import spacy

In [2]:
# Import the corpus
from nltk.corpus import brown, stopwords

brown.categories()

['adventure',
 'belles_lettres',
 'editorial',
 'fiction',
 'government',
 'hobbies',
 'humor',
 'learned',
 'lore',
 'mystery',
 'news',
 'religion',
 'reviews',
 'romance',
 'science_fiction']

In [3]:
# Create a DataFrame
df = pd.DataFrame()

In [4]:
# Populate the DataFrame with the fileids
df['ID'] = brown.fileids()

In [5]:
df['ID'].head()

0    ca01
1    ca02
2    ca03
3    ca04
4    ca05
Name: ID, dtype: object

In [6]:
# Populate the DataFrame with the text

text = [] # Create an empty list where the text can be stored
for each in df['ID']:
    txt = brown.words(fileids = [each]) # This returns the text based on the fileid 
    txt = [i.lower() for i in txt]
    txt = ' '.join(txt)
    text.append(txt) # Add the text to the list   

df['text'] = text # Create a new column in the DataFrame

In [7]:
df.head()

ID                                               text
0  ca01  the fulton county grand jury said friday an in...
1  ca02  austin , texas -- committee approval of gov. p...
2  ca03  several defendants in the summerdale police bu...
3  ca04  oslo the most positive element to emerge from ...
4  ca05  east providence should organize its civil defe...

In [8]:
# Populate the DataFrame with the genre of the text

categories = []
for each in df['ID']:
    cat = brown.categories(fileids = [each])
    cat = ' '.join(cat)
    categories.append(cat)

df['genre'] = categories

df.head() # Take a look at the DataFrame to see what we have.

ID                                               text genre
0  ca01  the fulton county grand jury said friday an in...  news
1  ca02  austin , texas -- committee approval of gov. p...  news
2  ca03  several defendants in the summerdale police bu...  news
3  ca04  oslo the most positive element to emerge from ...  news
4  ca05  east providence should organize its civil defe...  news

# Clustering

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(stop_words=stopwords.words('english'),
                            max_df=0.5,
                            min_df=0.1,
                            lowercase=False)

In [10]:
labels = df.genre
true_k = np.unique(labels).shape[0]

In [11]:
X = tfidf_vectorizer.fit_transform(df['text'])

In [12]:
from sklearn.decomposition import TruncatedSVD
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import Normalizer

svd = TruncatedSVD(true_k)
lsa = make_pipeline(svd,Normalizer(copy=False))

In [13]:
X = lsa.fit_transform(X)

In [14]:
from sklearn.cluster import KMeans

km = KMeans(n_clusters=true_k, init='k-means++', max_iter=100)
km.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=100,
    n_clusters=15, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [15]:
km.labels_

array([ 7,  9,  3, 14, 10,  7, 14, 14,  3,  8, 12, 12, 12, 12, 12,  4,  4,
        4,  3,  6,  4,  4,  4, 10,  3,  3,  3,  3,  4,  4,  8, 12,  4, 10,
        9,  7, 14, 12, 12, 12, 14, 14,  3,  3,  7, 14, 10, 14, 10, 14, 14,
        8, 10,  1, 14, 10,  1, 10,  3,  9, 14, 14, 14, 14, 14, 14, 14,  1,
       14, 14, 14,  1,  1,  1,  1,  1,  1,  1,  1, 10,  1, 10,  1,  1, 10,
        6, 14,  3, 11,  8,  8,  8,  8,  8,  8, 11,  8, 11, 11, 11,  2,  8,
       11,  8,  8, 10,  2, 14,  1,  4,  3,  2,  5, 12, 12,  6,  6, 13,  2,
        2,  2,  2,  6,  2,  2,  2,  1,  1,  9,  5,  9,  3,  3,  9,  3,  3,
        3,  3,  5,  3,  3,  1,  6, 11,  9,  2,  1, 11,  1,  6,  4,  2,  1,
        3, 14, 11,  6,  6,  6, 11,  6,  2,  6, 14,  6,  9,  2,  2,  6,  4,
        9,  4,  9,  9,  3,  2, 13,  8, 12, 11,  3,  8,  6,  1, 11,  6,  7,
        9, 11,  6, 14, 14, 13,  2,  1, 14,  6, 11,  7, 11,  1, 11, 13,  1,
       11,  6,  1, 11,  5, 14, 11, 11, 11, 11,  1,  6,  6, 11,  5,  1,  4,
        1,  1, 14, 10,  6

In [16]:
from sklearn import metrics

# metric of label giving ground truth
metrics.homogeneity_score(labels, km.labels_)

0.39266299793743681

In [17]:
metrics.completeness_score(labels, km.labels_)

0.3819388602370859

In [18]:
metrics.v_measure_score(labels, km.labels_)

0.38722669280174105

In [19]:
# rand index adjusted for change
metrics.adjusted_rand_score(labels, km.labels_)

0.16796493654276209

In [20]:
df['k'] = km.labels_

pd.crosstab(df['k'], df['genre'])

genre  adventure  belles_lettres  editorial  fiction  government  hobbies  \
k                                                                           
0             12               3          0       17           0        0   
1              0              13          3        1           0        3   
2              0               1          0        0           1        9   
3              0               0          1        0           8        9   
4              0               3          0        1           0        1   
5              0               2          0        0           1        3   
6              0              13          0        0           0        3   
7              0               1          1        0          12        0   
8              0               1          1        3           0        0   
9              0               2          1        0           1        3   
10             0               4          5        0           2        1   
11             0              21          0        0           0        0   
12             0               0          0        0           0        2   
13            17               2          0        7           0        1   
14             0               9         15        0           5        1   

genre  humor  learned  lore  mystery  news  religion  reviews  romance  \
k                                                                        
0          4        0     0       14     0         0        0       24   
1          4        9     5        0     0         0       11        0   
2          0        4     6        0     0         1        0        0   
3          0        5     3        0     9         0        1        0   
4          0        0     3        2     9         0        0        1   
5          0       33     0        0     0         0        0        0   
6          0        2    11        0     1         0        1        0   
7          0        4     1        0     3         0        0        0   
8          0        0     2        0     2        10        0        1   
9          0        6     6        0     2         0        0        0   
10         1        0     0        1     3         0        3        1   
11         0       12     7        0     0         6        0        0   
12         0        0     1        0     9         0        0        0   
13         0        1     1        7     0         0        0        2   
14         0        4     2        0     6         0        1        0   

genre  science_fiction  
k                       
0                    3  
1                    1  
2                    0  
3                    0  
4                    0  
5                    0  
6                    1  
7                    0  
8                    0  
9                    0  
10                   0  
11                   1  
12                   0  
13                   0  
14                   0

In [21]:
original_space_centroids = svd.inverse_transform(km.cluster_centers_)
order_centroids = original_space_centroids.argsort()[:, ::-1]
terms = tfidf_vectorizer.get_feature_names()

In [22]:
for i in range(true_k):
    print("Cluster %d:" % i, end='')
    for ind in order_centroids[i, :20]:
        print(' %s' % terms[ind], end='')
    print('\n')

Cluster 0: got eyes went thought knew door looked room going told something house face think felt asked woman mother let us

Cluster 1: music art love miss us young performance mother century form book though york experience stage american mr sense english often

Cluster 2: water surface used feet system small state head shown air area body earth light side pressure ground half size 10

Cluster 3: per 000 year industry cost program development system company million 1960 car market service equipment business cent government rate production

Cluster 4: mrs mr house president miss car white room home year family club woman police door wife school office children mother

Cluster 5: surface used system pressure shown number volume type water systems per information material method study range low small obtained possible

Cluster 6: south war north southern john river american company england wrote west english east states 000 century city year united york

Cluster 7: state states federal c

In [441]:
# Lets try the clusters using the DataFrame that does not inclue science fiction, humor, reviews or religion--
# df1

labels1 = df1.genre
true_k1 = np.unique(labels1).shape[0]

In [442]:
# Vectorize the text in the new DataFrame--use the vectorizer that we use in Unsupervised learning

X1 = vectorizer.fit_transform(df1['text'])

In [443]:
# Fit the kmeans cluster
fit_km = km.fit(X1)

In [451]:
# Assign the labels to a column in the new DataFrame

labels_km = km.labels_
df1['k'] = labels_km
centers1 = km.cluster_centers_

In [445]:
# metric of label giving ground truth
metrics.homogeneity_score(labels1, labels_km)

0.30832209604947969

In [446]:
metrics.completeness_score(labels1, labels_km)

0.28786868024763501

In [447]:
metrics.v_measure_score(labels1, labels_km)

0.29774454221584984

In [448]:
# rand index adjusted for change
metrics.adjusted_rand_score(labels1, labels_km)

0.10508326804476567

In [449]:
pd.crosstab(df1['k'], df1['genre'])

genre  adventure  belles_lettres  editorial  fiction  government  hobbies  \
k                                                                           
0              0               0          0        0           2        2   
1              0               2          1        0           0        3   
2              0               1          1        0          13       10   
3              0               5          0        0           0        1   
4              0               0          0        0           1        3   
5              0              13          1        1           1        2   
6              0              14          1        0           0        0   
7              0              11         20        0           5        2   
8              0              12          1        1           1        2   
9              1               2          0        2           0        1   
10            10               5          0       10           0        1   
11            18               5          0       14           0        2   
12             0               2          1        1           0        0   
13             0               0          1        0           6        4   
14             0               3          0        0           1        3   

genre  learned  lore  mystery  news  romance  
k                                             
0            6     0        0     0        0  
1            0     1        0    11        1  
2           17     3        0     4        0  
3            3     0        0     0        0  
4           22     0        0     0        0  
5            4     4        0     0        0  
6            2     5        0     0        2  
7            6     4        0    15        0  
8            3    11        0     1        0  
9            0     2        2     0        1  
10           2     7       13     7       15  
11           4     5        9     0       10  
12           0     0        0     0        0  
13           3     2        0     3        0  
14           8     4        0     3        0

In [455]:
# It appears that belles lettres are not being clustered together.  This could be because belles lettres is
# considered more of a writing style than a genre.  
# Make a DataFrame that drops belles lettres and rerun the clustering.

df_nobell = df1
df_nobell = df_nobell[df_nobell.genre != 'belles_lettres']

In [488]:
# Lets try the clusters using the DataFrame that does not inclue science fiction, humor, reviews or religion--
# df1

genre_nobell = df_nobell.genre
true_k_nobell = np.unique(genre_nobell).shape[0]

In [489]:
# Vectorize the text in the new DataFrame--use the vectorizer that we use in Unsupervised learning
X_nobell = vectorizer.fit_transform(df_nobell['text'])

# Fit the kmeans cluster
fit_nobell = km.fit(X_nobell)

# Assign the labels to a column in the new DataFrame
labels_nobell = km.labels_
df_nobell['k'] = labels_nobell

# Find the cluster centers
centers_nobell = km.cluster_centers_

/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [490]:
# metric of label giving ground truth
metrics.homogeneity_score(genre_nobell, labels_nobell)

0.35172659547154828

In [491]:
metrics.completeness_score(genre_nobell, labels_nobell)

0.32871192653160475

In [492]:
metrics.v_measure_score(genre_nobell, labels_nobell)

0.33983004509941395

In [493]:
# rand index adjusted for change
metrics.adjusted_rand_score(genre_nobell, labels_nobell)

0.13750381826279781

In [486]:
pd.crosstab(df_nobell['k'], df_nobell['genre'])

genre  adventure  editorial  fiction  government  hobbies  learned  lore  \
k                                                                          
0              0          0        0           0        0       10     0   
1             12          0        9           1        7        5     6   
2              0          1        2           0        1        3     3   
3              1          3        0           0        0        2     2   
4              2          1        4           0        3        1     5   
5              0         15        0           4        2        3     6   
6              0          0        0           5        6       10     1   
7              0          0        1           5        2        1     2   
8              0          0        0           1        9       20     2   
9              0          0        1           1        0        4     5   
10             0          2        0           6        3        2     3   
11             0          1        0           0        1       14     8   
12             0          1        0           7        1        2     2   
13             8          1        4           0        1        1     0   
14             6          2        8           0        0        2     3   

genre  mystery  news  romance  
k                              
0            0     0        0  
1            2     0        4  
2            3    11        1  
3            0     7        0  
4            0     7        3  
5            0     6        0  
6            0     2        0  
7            0     1        1  
8            0     0        0  
9            0     1        0  
10           0     8        0  
11           0     0        0  
12           0     1        0  
13          12     0        4  
14           7     0       16

In [487]:
# So dropping belles_letters did not work well either; I will keep belle_lettres out of
# the DataFrame and combine fiction, adventure, mystery, and romance.

df_last = df2
df_last = df_last[df_last.genre != 'belles_lettres']

In [494]:
genre_last = df_last.genre
true_k_last = np.unique(genre_last).shape[0]

In [497]:
# Vectorize the text in the new DataFrame--use the vectorizer that we use in Unsupervised learning
X_last = vectorizer.fit_transform(df_last['text'])

# Fit the kmeans cluster
fit_last = km.fit(X_last)

# Assign the labels to a column in the new DataFrame
labels_last = km.labels_
df_last['k'] = labels_last

# Find the cluster centers
centers_last = km.cluster_centers_

/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [499]:
# metric of label giving ground truth
metrics.homogeneity_score(genre_last, labels_last)

0.37746299541319928

In [501]:
metrics.completeness_score(genre_last, labels_last)

0.27747858178281504

In [502]:
metrics.v_measure_score(genre_last, labels_last)

0.31983889949744709

In [504]:
# rand index adjusted for change
metrics.adjusted_rand_score(genre_last, labels_last)

0.23010997503030736

In [505]:
pd.crosstab(df_last['k'], df_last['genre'])

genre  editorial  fiction  government  hobbies  learned  lore  news
k                                                                  
0              0        0           0        3        4     2     0
1              1        0          11       11       13     2     4
2              1        1           1        2        8     1     2
3              0        2           1        2        7     8     1
4              0        3           0        0        0     0     0
5              3        0           9        1        6     7     0
6              0        0           0        2        0     1     7
7              0        0           1        3       21     0     0
8              0       29           0        1        1     1     0
9              0       12           0        4        1     7     0
10             4        0           2        0        4     2    14
11            13        0           2        2        2     5    10
12             0        3           1        1        0     2     0
13             1        0           2        1       10     5     0
14             4       61           0        3        3     5     6

# Unsupervised Learning

Clustering worked better when I removed belles_lettres from the list of genres, so I am going to try unsupervised learning without it as well.

In [470]:
text_nb = df_nobell['text']
genre_nb = df_nobell['genre']

In [472]:
# Split the data into training and test sets

text_train_nb, text_test_nb, categories_train_nb, categories_test_nb = train_test_split(
    text_nb, genre_nb, test_size=0.25, random_state=42, stratify=genre_nb)

#Applying the vectorizer
train_tfidf_nb = vectorizer.fit_transform(text_train_nb)
print("Number of features: %d" % train_tfidf_nb.get_shape()[1])

test_tfidf_nb = vectorizer.transform(text_test_nb)
print("Number of features: %d" % test_tfidf_nb.get_shape()[1])

Number of features: 15598
Number of features: 15598


In [473]:
# Random Forest Classifier

fit_nb = rfc.fit(train_tfidf_nb, categories_train_nb)
predict_nb = rfc.predict(test_tfidf_nb)
score_nb = rfc.score(test_tfidf_nb, categories_test_nb)

print('Random Forest Classifier Score: {}'.format(score_nb))

Random Forest Classifier Score: 0.39361702127659576


In [474]:
# Cross-Validate

cvs_rfc_nb = cross_val_score(rfc, X_nobell, genre_nb)
print('Cross Validated Scores: {}'.format(cvs_rfc_nb))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_rfc_nb.mean(), cvs_rfc_nb.std() * 2))

Cross Validated Scores: [ 0.36220472  0.33070866  0.26229508]
Accuracy: 0.32 (+/- 0.08)


In [476]:
# Support Vector Machine

fit_nb_svc = svc.fit(train_tfidf_nb, categories_train_nb)
predict_nb_svc = svc.predict(test_tfidf_nb)
score_nb_svc = svc.score(test_tfidf_nb, categories_test_nb)

print('Support Vector Machine Classifier Score: {}'.format(score_nb_svc))

Support Vector Machine Classifier Score: 0.4787234042553192


In [477]:
# Cross-Validate

cvs_svc_nb = cross_val_score(svc, X_nobell, genre_nb)
print('Cross Validated Scores: {}'.format(cvs_svc_nb))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_svc_nb.mean(), cvs_svc_nb.std() * 2))

Cross Validated Scores: [ 0.51181102  0.45669291  0.36885246]
Accuracy: 0.45 (+/- 0.12)


In [479]:
# Logistic Regression

fit_nb_lr = lr.fit(train_tfidf_nb, categories_train_nb)
predict_nb_lr = lr.predict(test_tfidf_nb)
score_nb_lr = lr.score(test_tfidf_nb, categories_test_nb)

print('Logistic Regression Score: {}'.format(score_nb_lr))

Logistic Regression Score: 0.3723404255319149


In [508]:
#Gradient Boosting Classifier


fit_nb_gbc = gbc.fit(train_tfidf_nb.toarray(), categories_train_nb)
predict_nb_gbc = gbc.predict(test_tfidf_nb.toarray())
score_nb_gbc = gbc.score(test_tfidf_nb.toarray(), categories_test_nb)

print('Gradient Boosting Classifier Score: {}'.format(score_nb_gbc))

Gradient Boosting Classifier Score: 0.3829787234042553


In [ ]:
# No belles letters and combined fiction adventure mystery and romance.

In [511]:
text_last = df_last['text']
genre_last = df_last['genre']

In [512]:
# Split the data into training and test sets

text_train_last, text_test_last, categories_train_last, categories_test_last = train_test_split(
    text_last, genre_last, test_size=0.25, random_state=42, stratify=genre_nb)

#Applying the vectorizer
train_tfidf_last = vectorizer.fit_transform(text_train_last)
print("Number of features: %d" % train_tfidf_last.get_shape()[1])

test_tfidf_last = vectorizer.transform(text_test_last)
print("Number of features: %d" % test_tfidf_last.get_shape()[1])

Number of features: 15598
Number of features: 15598


In [514]:
# Random Forest Classifier

fit_last = rfc.fit(train_tfidf_last, categories_train_last)
predict_last = rfc.predict(test_tfidf_last)
score_last = rfc.score(test_tfidf_last, categories_test_last)

print('Random Forest Classifier Score: {}'.format(score_last))

Random Forest Classifier Score: 0.5212765957446809


In [516]:
# Support Vector Machine

fit_last_svc = svc.fit(train_tfidf_last, categories_train_last)
predict_last_svc = svc.predict(test_tfidf_last)
score_last_svc = svc.score(test_tfidf_last, categories_test_last)

print('Support Vector Machine Classifier Score: {}'.format(score_last_svc))

Support Vector Machine Classifier Score: 0.6276595744680851


In [517]:
# Logistic Regression

fit_last_lr = lr.fit(train_tfidf_last, categories_train_last)
predict_last_lr = lr.predict(test_tfidf_last)
score_last_lr = lr.score(test_tfidf_last, categories_test_last)

print('Logistic Regression Score: {}'.format(score_last_lr))

Logistic Regression Score: 0.5425531914893617


In [518]:
# KNN

fit_last_knn = knn.fit(train_tfidf_last, categories_train_last)
predict_last_knn = knn.predict(test_tfidf_last)
score_last_knn = knn.score(test_tfidf_last, categories_test_last)

print('Logistic Regression Score: {}'.format(score_last_knn))

Logistic Regression Score: 0.5531914893617021


Below is the unsupervised work that I did first.

In [23]:
from sklearn.model_selection import train_test_split

# Split the data into training and test sets

text_train, text_test, categories_train, categories_test = train_test_split(
    text, categories, test_size=0.25, random_state=42, stratify=categories)

In [24]:
pd.DataFrame(categories_test)[0].value_counts()

learned            20
belles_lettres     19
lore               12
news               11
hobbies             9
government          8
fiction             7
romance             7
editorial           7
adventure           7
mystery             6
religion            4
reviews             4
science_fiction     2
humor               2
Name: 0, dtype: int64

In [25]:
# Use sklearn.tfidfvectorizer to vectorize the data

vectorizer = TfidfVectorizer(max_df=0.5, # drop words that occur in more than half the paragraphs
                             min_df=2, # only use words that appear at least twice
                             stop_words='english', 
                             lowercase=False, #convert everything to lower case 
                             use_idf=True,#we definitely want to use inverse document frequencies in our weighting
                             norm=u'l2', #Applies a correction factor so that longer paragraphs and shorter paragraphs get treated equally
                             smooth_idf=True #Adds 1 to all document frequencies, as if an extra document existed that used every word once.  Prevents divide-by-zero errors
                            )


#Applying the vectorizer
train_tfidf=vectorizer.fit_transform(text_train)
print("Number of features: %d" % train_tfidf.get_shape()[1])

test_tfidf = vectorizer.transform(text_test)
print("Number of features: %d" % test_tfidf.get_shape()[1])

Number of features: 18959
Number of features: 18959


In [28]:
# Run a supervised model on the vectorized data

from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier()
fit = rfc.fit(train_tfidf, categories_train)
predict = rfc.predict(test_tfidf)
score = rfc.score(test_tfidf, categories_test)

In [29]:
print('Random Forest Classifier Score: {}'.format(score))

Random Forest Classifier Score: 0.44


In [31]:
# Cross-validate

vector_text = vectorizer.fit_transform(text)

from sklearn.model_selection import cross_val_score

cross_val_rfc = cross_val_score(rfc, vector_text, categories)

In [32]:
print('Cross Validated Scores: {}'.format(cross_val_rfc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_rfc.mean(), cross_val_rfc.std() * 2))

Cross Validated Scores: [ 0.31360947  0.25443787  0.29012346]
Accuracy: 0.29 (+/- 0.05)


These results are pretty awful.  I am going to analyze the data to see where the models are making mistakes and then adjust the data hopefully make the models more accurate.  

In [33]:
# Examine how many titles are in each genre

print(df['genre'].value_counts())

learned            80
belles_lettres     75
lore               48
news               44
hobbies            36
government         30
fiction            29
romance            29
adventure          29
editorial          27
mystery            24
religion           17
reviews            17
humor               9
science_fiction     6
Name: genre, dtype: int64


In [37]:
# We will drop the bottom four genres because there are not enough titles to adequately
# train the models to recognize them

df1 = df # create a new DataFrame


drop_list = ['science_fiction', 'humor', 'reviews', 'religion']
for each in drop_list:
    df1 = df1[df1.genre != each]

print(df1['genre'].value_counts())

learned           80
belles_lettres    75
lore              48
news              44
hobbies           36
government        30
fiction           29
romance           29
adventure         29
editorial         27
mystery           24
Name: genre, dtype: int64


In [39]:
# Create variables for the text and genre in the new DataFrame
text1 = df1['text']
categories1 = df1['genre']

In [40]:
# Split the data into training and test sets

text_train1, text_test1, categories_train1, categories_test1 = train_test_split(
    text1, categories1, test_size=0.25, random_state=42, stratify=categories1)

In [41]:
#Applying the vectorizer
train_tfidf1 = vectorizer.fit_transform(text_train1)
print("Number of features: %d" % train_tfidf1.get_shape()[1])

test_tfidf1 = vectorizer.transform(text_test1)
print("Number of features: %d" % test_tfidf1.get_shape()[1])

Number of features: 17616
Number of features: 17616


In [42]:
# Run a supervised model on the vectorized data

from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier()
fit1 = rfc.fit(train_tfidf1, categories_train1)
predict1 = rfc.predict(test_tfidf1)
score1 = rfc.score(test_tfidf1, categories_test1)

In [43]:
print('Random Forest Classifier Score: {}'.format(score1))

Random Forest Classifier Score: 0.40707964601769914


In [44]:
# Cross-validate

vector_text1 = vectorizer.fit_transform(text1)

cross_val_rfc1 = cross_val_score(rfc, vector_text1, categories1)

In [45]:
print('Cross Validated Scores: {}'.format(cross_val_rfc1))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_rfc1.mean(), cross_val_rfc1.std() * 2))

Cross Validated Scores: [ 0.34210526  0.30263158  0.27891156]
Accuracy: 0.31 (+/- 0.05)


It appears the model has gotten fractionally better, but it is still pretty awful.  After further examination of the clustered data above, it appears that adventure, fiction, mystery, and romance are frequently clustered together.  We will combine those categories under the genre 'fiction' and rerun the model and hope for increased accuracy.

In [304]:
# Create a new DataFrame

df2 = df1

# Rename all adventure, mystery, and romance titles
replace_list = ['adventure', 'mystery', 'romance']
for each in replace_list:
    df2 = df2.replace(each, 'fiction')

print(df2['genre'].value_counts())

fiction           111
learned            80
belles_lettres     75
lore               48
news               44
hobbies            36
government         30
editorial          27
Name: genre, dtype: int64


In [48]:
# Create variables for the text and genre in the new DataFrame
text2 = df2['text']
categories2 = df2['genre']

In [49]:
# Split the data into training and test sets

text_train2, text_test2, categories_train2, categories_test2 = train_test_split(
    text2, categories2, test_size=0.25, random_state=42, stratify=categories2)

In [50]:
#Applying the vectorizer
train_tfidf2 = vectorizer.fit_transform(text_train2)
print("Number of features: %d" % train_tfidf2.get_shape()[1])

test_tfidf2 = vectorizer.transform(text_test2)
print("Number of features: %d" % test_tfidf2.get_shape()[1])

Number of features: 17461
Number of features: 17461


In [351]:

print("Number of Rows: %d" % train_tfidf2.get_shape()[0])

print("Number of Rows: %d" % test_tfidf2.get_shape()[0])

Number of Rows: 338
Number of Rows: 113


In [51]:
# Run a supervised model on the vectorized data

fit2 = rfc.fit(train_tfidf2, categories_train2)
predict2 = rfc.predict(test_tfidf2)
score2 = rfc.score(test_tfidf2, categories_test2)

In [52]:
print('Random Forest Classifier Score: {}'.format(score2))

Random Forest Classifier Score: 0.48672566371681414


In [53]:
# Cross-validate

vector_text2 = vectorizer.fit_transform(text2)

cross_val_rfc2 = cross_val_score(rfc, vector_text2, categories2)

In [54]:
print('Cross Validated Scores: {}'.format(cross_val_rfc2))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_rfc2.mean(), cross_val_rfc2.std() * 2))

Cross Validated Scores: [ 0.53642384  0.43046358  0.4966443 ]
Accuracy: 0.49 (+/- 0.09)


In [519]:
# Support Vector Classifier

fit2_svc = svc.fit(train_tfidf2, categories_train2)
predict2_svc = svc.predict(test_tfidf2)
score2_svc = svc.score(test_tfidf2, categories_test2)

In [520]:
score2_svc

0.63716814159292035

In [509]:
# Gradient Boosting Classifier

fit_gbc2 = gbc.fit(train_tfidf2.toarray(), categories_train2)
predict_gbc2 = gbc.predict(test_tfidf2.toarray())
score_gbc2 = gbc.score(test_tfidf2.toarray(), categories_test2)

In [510]:
score_gbc2

0.53097345132743368

Combining adventure, fiction, mystery, and romance resulted in greatly increased model, however the model is still under fifty percent accurate.  More improvements need to be made.

In [55]:
# Lets try adding a stemmer and tokenizer to the vectorizer.

from nltk.stem import PorterStemmer
import string

stemmer = PorterStemmer()

def tokenize_and_stem(text):
    tokens = nltk.tokenize.word_tokenize(text)
    
    # strip out punctuation and make lowercase
    tokens = [token.lower().strip(string.punctuation)
             for token in tokens if token.isalnum()]
    
    # now stem the tokens
    tokens = [stemmer.stem(token) for token in tokens]
    
    return tokens

In [57]:
# Add the tokenize_and_stem function to the vectorizer

vectorizer2 = TfidfVectorizer(max_df=0.5, # drop words that occur in more than half the paragraphs
                             min_df=2, # only use words that appear at least twice
                             stop_words='english', 
                             lowercase=False, #convert everything to lower case 
                             use_idf=True,#we definitely want to use inverse document frequencies in our weighting
                             norm=u'l2', #Applies a correction factor so that longer paragraphs and shorter paragraphs get treated equally
                             smooth_idf=True, #Adds 1 to all document frequencies, as if an extra document existed that used every word once.  Prevents divide-by-zero errors
                             tokenizer = tokenize_and_stem
                            )

In [58]:
#Applying the vectorizer
train_tfidf3 = vectorizer2.fit_transform(text_train2)
print("Number of features: %d" % train_tfidf2.get_shape()[1])

test_tfidf3 = vectorizer2.transform(text_test2)
print("Number of features: %d" % test_tfidf2.get_shape()[1])

Number of features: 17461
Number of features: 17461


In [59]:
# Run a supervised model on the vectorized data

fit3 = rfc.fit(train_tfidf3, categories_train2)
predict3 = rfc.predict(test_tfidf3)
score3 = rfc.score(test_tfidf3, categories_test2)

In [60]:
print('Random Forest Classifier Score: {}'.format(score3))

Random Forest Classifier Score: 0.4778761061946903


In [61]:
# Cross-validate

vector_text3 = vectorizer2.fit_transform(text2)

cross_val_rfc3 = cross_val_score(rfc, vector_text3, categories2)

In [62]:
print('Cross Validated Scores: {}'.format(cross_val_rfc3))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_rfc3.mean(), cross_val_rfc3.std() * 2))

Cross Validated Scores: [ 0.53642384  0.40397351  0.48322148]
Accuracy: 0.47 (+/- 0.11)


In [521]:
# SVC

fit3_svc = svc.fit(train_tfidf3, categories_train2)
predict3_svc = svc.predict(test_tfidf3)
score3_svc = svc.score(test_tfidf3, categories_test2)

In [522]:
score3_svc

0.63716814159292035

Adding the tokenize_and_stem function to the vectorizer actually made the model worse.  Looks like I need to find another solution.

In [63]:
# Lets try different supervised learning models and see if we can find a more accurate model
# RandomForestClassifier

from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

In [64]:
# Run a support vector machine classifier

svc = SVC(kernel = 'linear')

fit_svc = svc.fit(train_tfidf2, categories_train2)
predict_svc = svc.predict(test_tfidf2)
score_svc = svc.score(test_tfidf2, categories_test2)

In [65]:
print('Support Vector Classifier Score: {}'.format(score_svc))

Support Vector Classifier Score: 0.6371681415929203


In [67]:
# Cross-validate

vector_text_svc = vectorizer.fit_transform(text2)

cross_val_svc = cross_val_score(svc, vector_text2, categories2)

In [68]:
print('Cross Validated Scores: {}'.format(cross_val_svc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_svc.mean(), cross_val_svc.std() * 2))

Cross Validated Scores: [ 0.62913907  0.53642384  0.55704698]
Accuracy: 0.57 (+/- 0.08)


In [69]:
# Run a Logistic Regression

lr = LogisticRegression()
fit_lr = lr.fit(train_tfidf2, categories_train2)
predict_lr = lr.predict(test_tfidf2)
score_lr = lr.score(test_tfidf2, categories_test2)

In [70]:
print('Logistic Regression Score: {}'.format(score_lr))

Logistic Regression Score: 0.4778761061946903


In [71]:
# Cross-validate

vector_text_lr = vectorizer.fit_transform(text2)

cross_val_lr = cross_val_score(lr, vector_text2, categories2)

In [72]:
print('Cross Validated Scores: {}'.format(cross_val_lr))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_lr.mean(), cross_val_lr.std() * 2))

Cross Validated Scores: [ 0.52980132  0.42384106  0.46308725]
Accuracy: 0.47 (+/- 0.09)


In [74]:
# Run a K Nearest Neighbors Classifier

from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier()
fit_knn = knn.fit(train_tfidf2, categories_train2)
predict_knn = knn.predict(test_tfidf2)
score_knn = knn.score(test_tfidf2, categories_test2)

In [75]:
print('KNN Classifier Score: {}'.format(score_knn))

KNN Classifier Score: 0.6460176991150443


In [76]:
# Cross-validate

vector_text_knn = vectorizer.fit_transform(text2)

cross_val_knn = cross_val_score(knn, vector_text2, categories2)

In [77]:
print('Cross Validated Scores: {}'.format(cross_val_knn))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_knn.mean(), cross_val_knn.std() * 2))

Cross Validated Scores: [ 0.60927152  0.52980132  0.52348993]
Accuracy: 0.55 (+/- 0.08)


In [435]:
# Gradient Boosting Classifier

# Create variables for the text and genre in the new DataFrame
text2_gbc = df2['text']
categories2_gbc = df2['genre']

# Split the data into training and test sets
text_train2_gbc, text_test2_gbc, categories_train2_gbc, categories_test2_gbc = train_test_split(
    text2_gbc, categories2_gbc, test_size=0.25, random_state=42, stratify=categories2_gbc)

#Applying the vectorizer
train_tfidf2_gbc = vectorizer.fit_transform(text_train2_gbc)
print("Number of features: %d" % train_tfidf2_gbc.get_shape()[1])

test_tfidf2_gbc = vectorizer.transform(text_test2_gbc)
print("Number of features: %d" % test_tfidf2_gbc.get_shape()[1])

# Run GBC
fit_gbc = gbc.fit(train_tfidf2_gbc.toarray(), categories_train2_gbc)
predict_gbc = gbc.predict(test_tfidf2_gbc.toarray())
score_gbc = gbc.score(test_tfidf2_gbc.toarray(), categories_test2_gbc)

print('Gradient Boosting Classifier Score: {}'.format(score_gbc))

Number of features: 17461
Number of features: 17461
Gradient Boosting Classifier Score: 0.5486725663716814


In [439]:
# Cross-validate

vector_text_gbc = vectorizer.fit_transform(text2_gbc)

cross_val_gbc = cross_val_score(gbc, vector_text_gbc.toarray(), categories2)

In [440]:
print('Cross Validated Scores: {}'.format(cross_val_gbc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cross_val_gbc.mean(), cross_val_gbc.std() * 2))

Cross Validated Scores: [ 0.53642384  0.53642384  0.41610738]
Accuracy: 0.50 (+/- 0.11)


The supervised model that outputted the best results using the vectorized data was the KNN Classifier with a score of 0.646 and a cross-validated score of 0.55 (+/- 0.08).  However, this is still a very poorly performing model.  In order to try to continue to improve the performance of the models, I am going to add more features using the bag-of-words method.  

In [221]:
# Utility function to create a list of the 10 most common words.
def bag_of_words(text):
    
    # Filter out punctuation and stop words.
    allwords = [token.lemma_
                for token in text
                if not token.is_punct
                and not token.is_stop]
    
    # Return the most common words.
    return [item[0] for item in Counter(allwords).most_common(11)]



In [151]:
common_words = []
for each in rar_list1:
    w = bag_of_words(each)
    for words in w:
        common_words.append(words)

In [154]:
common_words = set(common_words)
len(common_words)

3157

In [156]:
df2['genre'].value_counts()

fiction           111
learned            80
belles_lettres     75
lore               48
news               44
hobbies            36
government         30
editorial          27
Name: genre, dtype: int64

In [312]:
# Create DataFrames that isolate texts be genre

# fiction
df_fic = df2
df_fic = df_fic[df_fic.genre == 'fiction' ]

# learned
df_learn = df2
df_learn = df_learn[df_learn.genre == 'learned']

# belles_lettres
df_bell = df2
df_bell = df_bell[df_bell.genre == 'belles_lettres']

# lore
df_lore = df2
df_lore = df_lore[df_lore.genre == 'lore']

# news
df_news = df2
df_news = df_news[df_news.genre == 'news']

# hobbies
df_hob = df2
df_hob = df_hob[df_hob.genre == 'hobbies']

# government
df_gov = df2
df_gov = df_gov[df_gov.genre == 'government']

# editorial
df_ed = df2
df_ed = df_ed[df_ed.genre == 'editorial']

In [313]:
# Remove the `` from the text because it pops up as one of the most common words for each
# genre.
# Combine all the texts from each genre. Parse each newly created text.

df_fic['text'] = df_fic['text'].str.replace('``', '')
fiction = ''.join(df_fic['text'].tolist())
fiction = nlp(fiction)

df_learn['text'] = df_learn['text'].str.replace('``', '')
learned = ''.join(df_learn['text'].tolist())
learned = nlp(learned)

df_bell['text'] = df_bell['text'].str.replace('``', '')
belles = ''.join(df_bell['text'].tolist())
belles = nlp(belles)

#lore
df_lore['text'] = df_lore['text'].str.replace('``', '')
lore = ''.join(df_lore['text'].tolist())
lore = nlp(lore)

# news
df_news['text'] = df_news['text'].str.replace('``', '')
news = ''.join(df_news['text'].tolist())
news = nlp(news)

#hob
df_hob['text'] = df_hob['text'].str.replace('``', '')
hobbies = ''.join(df_hob['text'].tolist())
hobbies = nlp(hobbies)

df_gov['text'] = df_gov['text'].str.replace('``', '')
government= ''.join(df_gov['text'].tolist())
government = nlp(government)

df_ed['text'] = df_ed['text'].str.replace('``', '')
editorial = ''.join(df_ed['text'].tolist())
editorial = nlp(editorial)

/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/in

In [314]:
# Run all the parsed texts through the bag-of-words function to find the most common words
# for each genre.

fiction_bow = bag_of_words(fiction)
learned_bow = bag_of_words(learned)
belles_bow = bag_of_words(belles)
lore_bow = bag_of_words(lore)
news_bow = bag_of_words(news)
hobbies_bow = bag_of_words(hobbies)
government_bow = bag_of_words(government)
editorial_bow = bag_of_words(editorial)

In [315]:
# Combine the bags to create a set of unique words

common_words = set(fiction_bow + learned_bow + belles_bow + lore_bow + news_bow + hobbies_bow + government_bow + editorial_bow)

In [316]:
# Parse each of the texts individually

# Fiction
rar_list = []
for each in df_fic['text']:
    rar_list.append(each)
rar_list1 = []
for each in rar_list:
    t = nlp(each)
    rar_list1.append(t)
df_fic['text'] = rar_list1

# Learned
star_list = []
for each in df_learn['text']:
    star_list.append(each)
sun_list = []
for each in star_list:
    t = nlp(each)
    sun_list.append(t)
df_learn['text'] = sun_list

# Belles Lettres
star_list1 = []
for each in df_bell['text']:
    star_list1.append(each)
sun_list1 = []
for each in star_list1:
    t = nlp(each)
    sun_list1.append(t)
df_bell['text'] = sun_list1

# Lore
star_list2 = []
for each in df_lore['text']:
    star_list2.append(each)
sun_list2 = []
for each in star_list2:
    t = nlp(each)
    sun_list2.append(t)
df_lore['text'] = sun_list2

# News
star_list3 = []
for each in df_news['text']:
    star_list3.append(each)
sun_list3 = []
for each in star_list3:
    t = nlp(each)
    sun_list3.append(t)
df_news['text'] = sun_list3

# Hobbies
star_list4 = []
for each in df_hob['text']:
    star_list4.append(each)
sun_list4 = []
for each in star_list4:
    t = nlp(each)
    sun_list4.append(t)
df_hob['text'] = sun_list4

# Government
star_list5 = []
for each in df_gov['text']:
    star_list5.append(each)
sun_list5 = []
for each in star_list5:
    t = nlp(each)
    sun_list5.append(t)
df_gov['text'] = sun_list5

# Editorials
star_list6 = []
for each in df_ed['text']:
    star_list6.append(each)
sun_list6 = []
for each in star_list6:
    t = nlp(each)
    sun_list6.append(t)
df_ed['text'] = sun_list6

/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.

In [333]:
df_new = pd.DataFrame()

In [374]:
df_new = pd.concat([df_fic, df_learn, df_bell, df_lore, df_news, df_hob, df_gov, df_ed])

In [375]:
df_new['index'] = df_new.index

In [376]:
df_new = df_new.reset_index(drop=True)

In [377]:
df_new.head(20)

ID                                               text    genre   k  \
0   ck01  (thirty, -, three, scotty, did, not, go, back,...  fiction   0   
1   ck02  (where, their, sharp, edges, seemed, restless,...  fiction  13   
2   ck03  (mickie, sat, over, his, second, whisky, -, on...  fiction   0   
3   ck04  (the, bishop, looked, at, him, coldly, and, sa...  fiction   8   
4   ck05  (payne, dismounted, in, madison, place, and, h...  fiction   0   
5   ck06  (with, a, sneer, ,, the, man, spread, his, leg...  fiction   8   
6   ck07  (if, the, crummy, bastard, could, write, !, !,...  fiction   0   
7   ck08  (rousseau, is, so, persuasive, that, voltaire,...  fiction   1   
8   ck09  (it, was, the, first, time, any, of, us, had, ...  fiction  13   
9   ck10  (that, summer, the, gambling, houses, were, cl...  fiction   8   
10  ck11  (standing, in, the, shelter, of, the, tent, --...  fiction  13   
11  ck12  (she, was, a, child, too, much, a, part, of, h...  fiction   0   
12  ck13  (in, the, dim, underwater, light, they, dresse...  fiction   0   
13  ck14  (he, brought, with, him, a, mixture, of, myrrh...  fiction   0   
14  ck15  (beth, was, very, still, and, her, breath, cam...  fiction   0   
15  ck16  (the, red, glow, from, the, cove, had, died, o...  fiction  13   
16  ck17  (burly, leathered, men, and, wrinkled, women, ...  fiction   0   
17  ck18  (she, was, getting, real, dramatic, ., i, 'd, ...  fiction   0   
18  ck19  (there, was, one, fact, which, rector, could, ...  fiction   0   
19  ck20  (she, concluded, by, asking, him, to, name, an...  fiction   0   

    index  
0     374  
1     375  
2     376  
3     377  
4     378  
5     379  
6     380  
7     381  
8     382  
9     383  
10    384  
11    385  
12    386  
13    387  
14    388  
15    389  
16    390  
17    391  
18    392  
19    393

In [378]:
# Creates a data frame with features for each word in our common word set.
# Each value is the count of the times the word appears in each sentence.
def bow_features(common_words):
    
    # Scaffold the data frame and initialize counts to zero.
    df_bow = pd.DataFrame(columns=common_words)
    df_bow['text'] = df_new['text']
    df_bow.loc[:, common_words] = 0
    
    # Process each row, counting the occurrence of words in each sentence.
    for i, corp in enumerate(df_bow['text']):
        
        # Convert the sentence to lemmas, then filter out punctuation,
        # stop words, and uncommon words.
        words = [token.lemma_
                 for token in corp
                 if (
                     not token.is_punct
                     and not token.is_stop
                     and token.lemma_ in common_words
                 )]
        
        # Populate the row with word counts.
        for word in words:
            df_bow.loc[i, word] += 1
        
        # This counter is just to make sure the kernel didn't hang.
        if i % 500 == 0:
            print("Processing row {}".format(i))
            
    return df_bow


In [379]:
# Create our data frame with features. This can take a while to run.
word_counts = bow_features(common_words)

Processing row 0


In [380]:
word_counts.head()

year mr not world  1 think point new mrs president  \
0    0  6   5     0  0     6     0   1   1         0   
1    0  0   6     0  0     4     4   3   0         0   
2    1  2  12     1  0     4     1   2   3         9   
3    8  0   5     0  0     3     0   4   2         0   
4    0  1   3     2  0     2     1   2   4         1   

                         ...                         place school have system  \
0                        ...                             0      6   27      0   
1                        ...                             3      0   13      0   
2                        ...                             1      1   14      0   
3                        ...                             6      1    2      1   
4                        ...                             1      0   10      0   

  like look come people  2                                               text  
0    6   10    5      0  0  (thirty, -, three, scotty, did, not, go, back,...  
1    9    3    4      2  0  (where, their, sharp, edges, seemed, restless,...  
2    9    6    7      1  1  (mickie, sat, over, his, second, whisky, -, on...  
3    7   10    3      6  0  (the, bishop, looked, at, him, coldly, and, sa...  
4    9    4    2      2  0  (payne, dismounted, in, madison, place, and, h...  

[5 rows x 44 columns]

In [381]:
word_counts['index'] = df_new['index']

In [382]:
word_counts['genre'] = df_new['genre']

In [384]:
text_vec = vectorizer.fit_transform(df2['text'])

In [390]:
text_vec

<451x20647 sparse matrix of type '<class 'numpy.float64'>'
	with 240774 stored elements in Compressed Sparse Row format>

In [396]:
print(text_vec.toarray())

[[ 0.          0.          0.         ...,  0.          0.          0.        ]
 [ 0.          0.1040955   0.         ...,  0.          0.          0.        ]
 [ 0.          0.03386396  0.         ...,  0.          0.          0.        ]
 ..., 
 [ 0.          0.          0.         ...,  0.          0.          0.        ]
 [ 0.          0.          0.         ...,  0.          0.          0.        ]
 [ 0.          0.          0.         ...,  0.          0.          0.        ]]


In [400]:
list_ = ['a'+str(i) for i in range(20647)]
print(list_[0])

a0


In [401]:
df_vec = pd.DataFrame(text_vec.toarray(), columns = list_)

In [402]:
df_vec.head()

a0        a1   a2   a3   a4   a5   a6   a7   a8        a9   ...    a20637  \
0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.010579   ...       0.0   
1  0.0  0.104096  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.011233   ...       0.0   
2  0.0  0.033864  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.076740   ...       0.0   
3  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.000000   ...       0.0   
4  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.025149   ...       0.0   

   a20638  a20639  a20640  a20641  a20642  a20643  a20644  a20645  a20646  
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
3     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  

[5 rows x 20647 columns]

In [403]:
df_vec['index'] = df_vec.index

In [404]:
len(df_vec)

451

In [405]:
word_counts = pd.merge(df_vec, word_counts, how='inner', on='index')
word_counts.head()

a0        a1   a2   a3   a4   a5   a6   a7   a8        a9  ...    school  \
0  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.010579  ...         5   
1  0.0  0.104096  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.011233  ...        19   
2  0.0  0.033864  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.076740  ...         9   
3  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.000000  ...         0   
4  0.0  0.000000  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.025149  ...         0   

   have  system  like  look  come  people  2  \
0    19       2     2     0     0       0  3   
1    11       0     3     0     1       3  0   
2     9       3     0     0     0       3  1   
3    10       1     0     1     2       0  0   
4     8       2     0     0     7       1  0   

                                                text  genre  
0  (the, fulton, county, grand, jury, said, frida...   news  
1  (austin, ,, texas, --, committee, approval, of...   news  
2  (several, defendants, in, the, summerdale, pol...   news  
3  (oslo, the, most, positive, element, to, emerg...   news  
4  (east, providence, should, organize, its, civi...   news  

[5 rows x 20693 columns]

In [525]:
# Create target and predictor variables; normalize the predictor variables

from sklearn.preprocessing import normalize

features = word_counts.drop(['text', 'genre', 'index'], axis=1)
features = normalize(features)
target_var = word_counts['genre']


# Split the data into training and test sets

feat_train, feat_test, genre_train, genre_test = train_test_split(
    features, target_var, test_size=0.25, random_state=0, stratify=target_var)

In [526]:
# Run the supervised models on the DataFrame---
# Random Forest, Support Vector Machine, Logistic Regression, KNN

# Random Forest
fit = rfc.fit(feat_train, genre_train)
predict = rfc.predict(feat_test)
score = rfc.score(feat_test, genre_test)

print('Random Forest Classifier Score: {}'.format(score))

Random Forest Classifier Score: 0.47572815533980584


In [527]:
# Cross-validate

cvs_rfc = cross_val_score(rfc, features, target_var)
print('Cross Validated Scores: {}'.format(cvs_rfc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_rfc.mean(), cvs_rfc.std() * 2))

Cross Validated Scores: [ 0.34057971  0.4057971   0.31851852]
Accuracy: 0.35 (+/- 0.07)


In [528]:
# Support Vector Classifier
fit = svc.fit(feat_train, genre_train)
predict = svc.predict(feat_test)
score = svc.score(feat_test, genre_test)

print('Support Vector Classifier Score: {}'.format(score))

Support Vector Classifier Score: 0.5922330097087378


In [529]:
# Cross-validate
cvs_svc = cross_val_score(svc, features, target_var)
print('Cross Validated Scores: {}'.format(cvs_svc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_svc.mean(), cvs_svc.std() * 2))

Cross Validated Scores: [ 0.5942029   0.46376812  0.47407407]
Accuracy: 0.51 (+/- 0.12)


In [530]:
# Logistic Regression
fit = lr.fit(feat_train, genre_train)
predict = lr.predict(feat_test)
score = lr.score(feat_test, genre_test)

print('Logistic Regression Score: {}'.format(score))

Logistic Regression Score: 0.6116504854368932


In [531]:
# Cross-validate
cvs_lr = cross_val_score(lr, features, target_var)
print('Cross Validated Scores: {}'.format(cvs_lr))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_lr.mean(), cvs_lr.std() * 2))

Cross Validated Scores: [ 0.56521739  0.42753623  0.47407407]
Accuracy: 0.49 (+/- 0.11)


In [532]:
# KNN
fit = knn.fit(feat_train, genre_train)
predict = knn.predict(feat_test)
score = knn.score(feat_test, genre_test)

print('KNN Classifier Score: {}'.format(score))

KNN Classifier Score: 0.5145631067961165


In [533]:
# Cross-validate
cvs_knn = cross_val_score(knn, features, target_var)
print('Cross Validated Scores: {}'.format(cvs_knn))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_knn.mean(), cvs_knn.std() * 2))

Cross Validated Scores: [ 0.54347826  0.39855072  0.42222222]
Accuracy: 0.45 (+/- 0.13)


In [534]:
# Gradient Boosting Classifier
from sklearn.ensemble import GradientBoostingClassifier

gbc = GradientBoostingClassifier()
fit = gbc.fit(feat_train, genre_train)
predict = gbc.predict(feat_test)
score = gbc.score(feat_test, genre_test)

print('Gradient Boosting Classifier Score: {}'.format(score))

Gradient Boosting Classifier Score: 0.5631067961165048


In [535]:
# Cross-validate
cvs_gbc = cross_val_score(gbc, features, target_var)
print('Cross Validated Scores: {}'.format(cvs_gbc))
print("Accuracy: %0.2f (+/- %0.2f)" % (cvs_gbc.mean(), cvs_gbc.std() * 2))

Cross Validated Scores: [ 0.38405797  0.47826087  0.44444444]
Accuracy: 0.44 (+/- 0.08)
